### PR TITLE
extend itip message format

### PR DESCRIPTION
**Pull request by elliefm**
This extends `imip_send()` to include the JMAP id (as `id:`) and ITIP method (as `method:`) fields in IMIP notifications.  I did this by adding these as individual new fields, though...

Michael said:
> Actually, it would be really useful to just have the same thing we have when we include `calendarEvents` in Email/get, which comes from `jmap_calendar_events_from_msg`.

I had a look at this as an option, and both things end up in `jmapical_tojmap_all()` anyway.  So it's not clear to me why `jmap_calendar_events_from_msg()` would be producing additional fields beyond what `imip_send()` already was?

I'd like to add some tests, I'm concerned that recurid seems to always be NULL in what I've looked at so far... but maybe that's expected for IMIP, I dunno.  I'd appreciate some guidance in identifying which tests I should copy and modify, there's a lot with "itip" or "imip" in the name already, but they all pass, which means none of them cared much about the structure that I've changed.

There's also a bunch of junk debug logging in here, which I'll rebase out once I don't need it anymore.

### CODE REVIEW

**Review by ksmurchison — APPROVED** [review]
LGTM after debug cruft is removed

**Comment by elliefm** [comment]
@ksmurchison Thanks.  Any comment on `recurid` seeming to always be NULL?  I can't tell if that's expected, or indicative that I'm doing something subtly wrong.  If I knew what I was doing I'd just write some tests, but I don't, and don't know what tests to write.

**Comment by ksmurchison** [comment]
@elliefm The reason the `recurid` is always NULL is that sched_data->itip is actually a VCALENDAR component, not a VEVENT component, so it will never find a RECURRENCE-ID property.  `icalcomponent_get_uid()` has some magic which will dive down inside VCALENDAR to find a VEVENT, VTODO, etc, but `icalcomponent_get_first_property()` only operates on the component given to it.  You're going to want to do something like 

```
prop = icalcomponent_get_first_property(icalcomponent_get_first_real_component(sched_data->itip),
                                        ICAL_RECURRENCEID_PROPERTY)
```
As far as creating a test, one of the few ways that an iTIP message would have a recurrence-id is if an attendee is invited to only a single instance of a recurring event.

**Comment by elliefm** [comment]
Thanks

Looks like `JMAPCalendars.calendarevent_set_standalone_itip` already deals with itip notifications for events where the participant is only invited to a single recurrence, so I'll tinker with that to have it verify the `id` field at the same time.

**Comment by elliefm** [comment]
I was seeing failures from `JMAPCalendars.calendarevent_changes_issue2558` an hour or so ago which seemed to be coming from this branch, but now that I start trying to investigate, they've stopped happening.  Curious.  Will have a look again on Wednesday.

I've extended all the tests I could find that expected to get ITIP notifications, and those now all verify that the new `id` and `method` fields contain sensible values.

**Review by ksmurchison — APPROVED** [review]
This all looks good, once the debugging cruft is removed.

**Comment by elliefm** [comment]
@rsto This is now ready for a proper review :)

File: cassandane/tiny-tests/JMAPCalendars/calendarevent-encode-imip-uri [hunk 1]
```diff
@@ -69,6 +71,9 @@ EOF
     my ($imipnotif) = grep { $_->{METHOD} eq 'imip' } @$data;
     my $payload = decode_json($imipnotif->{MESSAGE});
     $self->assert_str_equals('plus+uri@example.com', $payload->{recipient});
+    my $expect_id = "E-$uid";
```
Comment (rsto) [inline cassandane/tiny-tests/JMAPCalendars/calendarevent-encode-imip-uri:74]: Please replace the hardcoded event id strings with `$self->encode_eventid($uid, $recurid)`. The recurid argument is optional.

**Review by rsto — CHANGES_REQUESTED** [review]

File: cassandane/tiny-tests/JMAPCalendars/calendarevent-encode-imip-uri (hunk 1 above)
Comment (elliefm) [reply cassandane/tiny-tests/JMAPCalendars/calendarevent-encode-imip-uri:74]: Oh, great tip, thanks!

Comment (elliefm) [reply cassandane/tiny-tests/JMAPCalendars/calendarevent-encode-imip-uri:74]:
```
    my $expect_id = $self->encode_eventid($uid);
    xlog "XXX expect id $uid => $expect_id";
```

```
[4318] =====> Cyrus::JMAPCalendars::test_calendarevent_encode_imip_uri[75] XXX expect id event1uid => ERB-event1uid-Q2Fzc2FuZGFuZTo6Q3lydXM6OkpNQVBDYWxlbmRhcnM9SEFTSCgweDU2MWRlMjhhNjczOCk
```

Weird, `encode_eventid($uid)` is adding a load of (base64?) junk as a recurid when the recurid argument is omitted.

It's strange because I see lots of other tests using `encode_eventid($uid)` without an argument, and they're not failing.  I guess maybe they're comparing eventids where both were generated by Cassandane, so the difference from Cyrus's encoder isn't noticed?  Hmm.

Comment (elliefm) [reply cassandane/tiny-tests/JMAPCalendars/calendarevent-encode-imip-uri:74]: Oh I see what's wrong, it's not an object method, it's just a normal function.  So I shouldn't be calling it with `$self->` :)

File: imap/http_caldav_sched.c [hunk 2]
```diff
@@ -632,18 +632,45 @@ static int imip_send(const char *userid, struct sched_data *sched_data,
...
     }
+
+    struct jmap_caleventid eid = { 0 };
+    icalcomponent *comp = NULL;
+    icalproperty *prop = NULL;
+    struct buf buf = BUF_INITIALIZER;
+
+    /* if there's going to be a recurrence-id, it will be a layer down */
+    comp = icalcomponent_get_first_real_component(sched_data->itip);
+    if (comp) {
```
Comment (dilyanpalauzov) [inline imap/http_caldav_sched.c:643]: The construct
```c
icalcomponent *comp = NULL;
comp = icalcomponent_get_first_real_component(sched_data->itip);
```
leads to dead initialization warning, e.g. in clang static analyzer (kind of useless assignment).  The more concise
```c
icalcomponent *comp = icalcomponent_get_first_real_component(sched_data->itip);
```
produces no such warning.

Comment (elliefm) [reply imap/http_caldav_sched.c:643]: That's interesting, can you paste the warning?  I don't have a recent clang handy

Comment (dilyanpalauzov) [reply imap/http_caldav_sched.c:643]: https://mail.aegee.org/dpa/scan99/2022-08-16-122910-1822492-1/

**Comment by elliefm** [comment]
@rsto I've made the changes you requested, and I've also fixed the dead initialisation that @dilyanpalauzov found.  Can you re-review please?

File: imap/http_caldav_sched.c (hunk 2 above)
Comment (rsto) [reply imap/http_caldav_sched.c:643]: I would prefer just as @dilyanpalauzov to use single-line declaration and initialization.

**Review by rsto — CHANGES_REQUESTED** [review]

File: imap/http_caldav_sched.c (hunk 2 above)
Comment (elliefm) [reply imap/http_caldav_sched.c:642]: done

**Review by rsto — APPROVED** [review]
